### PR TITLE
[eas-cli] fingerprint:compare

### PR DESCRIPTION
**Pull request by quinlanj**
# Why

Added a new command `eas fingerprint:compare` to help developers understand differences between their local project and a build's fingerprint. This helps diagnose issues where builds are not matching the expected configuration.

For now, this is a first-pass attempt. I'd like to focus on outputting an easy to understand diff for fingerprint, so:
- `eas fingerprint:compare` is hidden for now
- We only support passing in `buildId` or choosing a build from the cli menu, for now

Based on https://www.notion.so/expo/Soft-fingerprinting-116e5b5735248071951cddcab32ae43d?pvs=4#116e5b57352480c6a376e17971f48e8a

# Output from test project with differing fingerprint
```
MacBook-Pro-8:staging-sdk-52 quin$ EXPO_STAGING=1 ~/Documents/eas-cli/packages/eas-cli/bin/run fingerprint:compare
✔ Which build do you want to compare? › 🍏 Started at: 11/15/2024, 1:19:40 PM, Status: finished, Id: 1aa82a8d-6fed-4c4c-ab93-ef07d0c7ca97
Comparing fingerprints of the current project and build 1aa82a8d-6fed-4c4c-ab93-ef07d0c7ca97…
✔ Computed project fingerprint
🔄 Project fingerprint differs from build

📁 Paths with native dependencies:
    new directory::  node_modules/expo-eas-client/ios
    new directory::  node_modules/expo-json-utils/ios
    new directory::  node_modules/expo-manifests/ios
    new directory::  node_modules/expo-structured-headers/ios
    new directory::  node_modules/expo-updates-interface/ios
    new directory::  node_modules/expo-updates/ios

📝 modified content: Expo autolinking config (iOS)
@@ -81,0 +81,19 @@
     },
     {
+      "packageName": "expo-eas-client",
+      "packageVersion": "0.13.1",
+      "pods": [
+        {
+          "podName": "EASClient",
+          "podspecDir": "node_modules/expo-eas-client/ios"
+        }
+      ],
+      "swiftModuleNames": [
+        "EASClient"
+      ],
+      "modules": [
+        "EASClientModule"
+      ],
+      "appDelegateSubscribers": [],
+      "reactDelegateHandlers": [],
+      "debugOnly": false
+    },
+    {
       "packageName": "expo-file-system",
       "packageVersion": "18.0.3",
@@ -141,0 +160,17 @@
     },
     {
+      "packageName": "expo-json-utils",
+      "packageVersion": "0.14.0",
+      "pods": [
+        {
+          "podName": "EXJSONUtils",
+          "podspecDir": "node_modules/expo-json-utils/ios"
+        }
+      ],
+      "swiftModuleNames": [
+        "EXJSONUtils"
+      ],
+      "modules": [],
+      "appDelegateSubscribers": [],
+      "reactDelegateHandlers": [],
+      "debugOnly": false
+    },
+    {
       "packageName": "expo-keep-awake",
       "packageVersion": "14.0.1",
@@ -181,0 +217,17 @@
     },
     {
+      "packageName": "expo-manifests",
+      "packageVersion": "0.15.4",
+      "pods": [
+        {
+          "podName": "EXManifests",
+          "podspecDir": "node_modules/expo-manifests/ios"
+        }
+      ],
+      "swiftModuleNames": [
+        "EXManifests"
+      ],
+      "modules": [],
+      "appDelegateSubscribers": [],
+      "reactDelegateHandlers": [],
+      "debugOnly": false
+    },
+    {
       "packageName": "expo-modules-core",
       "packageVersion": "2.0.3",
@@ -240,0 +293,17 @@
     },
     {
+      "packageName": "expo-structured-headers",
+      "packageVersion": "4.0.0",
+      "pods": [
+        {
+          "podName": "EXStructuredHeaders",
+          "podspecDir": "node_modules/expo-structured-headers/ios"
+        }
+      ],
+      "swiftModuleNames": [
+        "EXStructuredHeaders"
+      ],
+      "modules": [],
+      "appDelegateSubscribers": [],
+      "reactDelegateHandlers": [],
+      "debugOnly": false
+    },
+    {
       "packageName": "expo-symbols",
       "packageVersion": "0.2.0",
@@ -278,0 +348,38 @@
     },
     {
+      "packageName": "expo-updates",
+      "packageVersion": "0.26.7",
+      "pods": [
+        {
+          "podName": "EXUpdates",
+          "podspecDir": "node_modules/expo-updates/ios"
+        }
+      ],
+      "swiftModuleNames": [
+        "EXUpdates"
+      ],
+      "modules": [
+        "UpdatesModule"
+      ],
+      "appDelegateSubscribers": [],
+      "reactDelegateHandlers": [
+        "ExpoUpdatesReactDelegateHandler"
+      ],
+      "debugOnly": false
+    },
+    {
+      "packageName": "expo-updates-interface",
+      "packageVersion": "1.0.0",
+      "pods": [
+        {
+          "podName": "EXUpdatesInterface",
+          "podspecDir": "node_modules/expo-updates-interface/ios"
+        }
+      ],
+      "swiftModuleNames": [
+        "EXUpdatesInterface"
+      ],
+      "modules": [],
+      "appDelegateSubscribers": [],
+      "reactDelegateHandlers": [],
+      "debugOnly": false
+    },
+    {
       "packageName": "expo-web-browser",
       "packageVersion": "14.0.1",

📝 modified content: Expo config
@@ -47,0 +47,3 @@
     ]
   ],
+  "runtimeVersion": {
+    "policy": "appVersion"
+  },
   "scheme": "myapp",
   "sdkVersion": "52.0.0",
@@ -50,0 +53,3 @@
   "slug": "staging-sdk-52",
+  "updates": {
+    "url": "https://staging-u.expo.dev/4a0663e3-6834-4a9b-8ac1-f82c0d29b7f9"
+  },
   "userInterfaceStyle": "automatic",
   "version": "1.0.0",

```

# How

- Added new GraphQL types and queries to support fetching build fingerprints
- Created a new command `eas fingerprint compare` that:
  - Fetches a build's fingerprint
  - Generates a fingerprint of the local project
  - Shows a detailed diff of the differences between them
- Added utilities for displaying diffs in a git-like format
- Added the `diff` package for computing text differences

# Test Plan

- [ ] new tests pass

### CODE REVIEW

**Comment by quinlanj** [comment]
* **#2736** <a href="https://app.graphite.dev/github/pr/expo/eas-cli/2736?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/expo/eas-cli/2736?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/eas-cli/src/commands/fingerprint/compare.ts [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  contentsId: string;
+}): void {
+  Log.newLine();
+  const prettyContentId = PRETTY_CONTENT_ID[contentsId] ?? contentsId;
+  if (op === 'added') {
+    Log.log(`${chalk.dim('📝 new content')}: ${prettyContentId}`);
+  } else if (op === 'removed') {
+    Log.log(`${chalk.dim('📝 removed content')}: ${prettyContentId}`);
+  } else if (op === 'changed') {
+    Log.log(`${chalk.dim('📝 modified content')}: ${prettyContentId}`);
```
Comment (brentvatne) [inline packages/eas-cli/src/commands/fingerprint/compare.ts:284]:
```suggestion
    Log.log(`${chalk.dim('📝 Modified content')}: ${prettyContentId}`);
```

File: packages/eas-cli/src/commands/fingerprint/compare.ts [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+  contentsBefore?: string | Buffer;
+  contentsAfter?: string | Buffer;
+  contentsId: string;
+}): void {
+  Log.newLine();
+  const prettyContentId = PRETTY_CONTENT_ID[contentsId] ?? contentsId;
+  if (op === 'added') {
+    Log.log(`${chalk.dim('📝 new content')}: ${prettyContentId}`);
+  } else if (op === 'removed') {
+    Log.log(`${chalk.dim('📝 removed content')}: ${prettyContentId}`);
```
Comment (brentvatne) [inline packages/eas-cli/src/commands/fingerprint/compare.ts:282]:
```suggestion
    Log.log(`${chalk.dim('📝 Removed content')}: ${prettyContentId}`);
```

File: packages/eas-cli/src/commands/fingerprint/compare.ts [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+  op: 'added' | 'removed' | 'changed';
+  sourceType: 'contents';
+  contentsBefore?: string | Buffer;
+  contentsAfter?: string | Buffer;
+  contentsId: string;
+}): void {
+  Log.newLine();
+  const prettyContentId = PRETTY_CONTENT_ID[contentsId] ?? contentsId;
+  if (op === 'added') {
+    Log.log(`${chalk.dim('📝 new content')}: ${prettyContentId}`);
```
Comment (brentvatne) [inline packages/eas-cli/src/commands/fingerprint/compare.ts:280]:
```suggestion
    Log.log(`${chalk.dim('📝 New content')}: ${prettyContentId}`);
```

File: packages/eas-cli/src/commands/fingerprint/compare.ts [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+  sourceType,
+  filePath,
+}: {
+  op: 'added' | 'removed' | 'changed';
+  sourceType: 'dir' | 'file';
+  filePath: string;
+}): FormatFieldsItem {
+  if (sourceType === 'dir') {
+    if (op === 'added') {
+      return { label: 'new directory:', value: filePath };
```
Comment (brentvatne) [inline packages/eas-cli/src/commands/fingerprint/compare.ts:240]: in the example output from the op it looks like there were two `:` in the output - "new directory::` - it'd be good to double check that so we only have one.

**Review by brentvatne — CHANGES_REQUESTED** [review]
my preference right now is to only show the diff (let's ensure it's also colored red/green), and keep the additional analysis that we provide behind an env var to iterate on. `EAS_ENABLE_FINGERPRINT_SUMMARY=1` could toggle this on for now, and we can add that by default later

File: packages/eas-cli/src/commands/fingerprint/compare.ts [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+      let sourceType;
+      if (diff.op === 'added') {
+        sourceType = diff.addedSource.type;
+      } else if (diff.op === 'removed') {
+        sourceType = diff.removedSource.type;
+      } else if (diff.op === 'changed') {
+        sourceType = diff.beforeSource.type;
+      }
+      return sourceType === 'dir' || sourceType === 'file';
+    });
```
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/eas-cli/src/commands/fingerprint/compare.ts:120]: https://app.graphite.dev/github/pr/expo/eas-build/473/build-tools-Remove-fallback-unversioned-fingerprint-diff

**Comment by wschurman** [comment]
I guess we could do the diff parsing locally in eas-cli for a bit and then gradually move over to a versioned diff parser/printer? (like we've done for diffing itself in eas-build)

File: packages/eas-cli/src/commands/fingerprint/compare.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (wschurman) [reply packages/eas-cli/src/commands/fingerprint/compare.ts:120]: Chatted about this offline: decided to keep pretty diff code local in eas-cli here for now while we add a versioned pretty diff command/function, then remove the local code here afterwards.

File: packages/eas-cli/src/utils/fingerprintCli.ts [hunk 6]
```diff
@@ -13,14 +14,30 @@ export type FingerprintOptions = {
...
+  fingerprint2: Fingerprint
+): FingerprintDiffItem[] | null {
+  // @expo/fingerprint is exported in the expo package for SDK 52+
+  const fingerprintPath = silentResolveFrom(projectDir, 'expo/fingerprint');
+  if (!fingerprintPath) {
+    return null;
+  }
+
+  const Fingerprint = require(fingerprintPath);
+  return Fingerprint.diffFingerprints(fingerprint1, fingerprint2);
```
Comment (wschurman) [inline packages/eas-cli/src/utils/fingerprintCli.ts:29]: add a test in `@expo/fingerprint` package to ensure that that this function signature remains stable, similar to what we have for `Fingerprint.createFingerprintAsync`.

File: packages/eas-cli/tsconfig.json [hunk 7]
```diff
@@ -10,7 +10,7 @@
     "noUnusedLocals": true,
     "noUnusedParameters": true,
     "strict": true,
-    "target": "ES2021",
```
Comment (wschurman) [inline packages/eas-cli/tsconfig.json:13]: we should probably extend one of the tsconfig bases (https://github.com/tsconfig/bases#centralized-recommendations-for-tsconfig-bases) to ensure this target is compatible with the minumum node version that we support for eas-cli.

That being said, I checked and we support node >=18 which uses es2022 in the base: https://www.npmjs.com/package/@tsconfig/node18

**Review by wschurman — APPROVED** [review]

File: packages/eas-cli/src/utils/fingerprintCli.ts (hunk 6 above)
Comment (quinlanj) [reply packages/eas-cli/src/utils/fingerprintCli.ts:29]: https://github.com/expo/expo/pull/33410

File: packages/eas-cli/tsconfig.json (hunk 7 above)
Comment (quinlanj) [reply packages/eas-cli/tsconfig.json:13]: https://github.com/expo/eas-cli/pull/2739

**Pull request merged**